Updated sim mode docs (#641)
cylc/cylc-doc: src/user-guide/running-workflows/simulation-modes.rst

run length is computed by dividing the time limit by the speedup factor.

Simulated Failure
^^^^^^^^^^^^^^^^^

> By default all tasks succeed, sending all task outputs.

Reviewer: That would include `failed` and `submit-failed`?

I'm guessing that by default simulation and dummy mode only send:

* submitted
* started
* succeeded

(and no custom outputs)

Reviewer: From testing, it would appear that `itask.tdef.rtconfig['outputs'].values()` is empty (possibly unset by the simulation config?), so it looks like simulation mode only sets `succeeded` here.

Reviewer: ```diff
diff --git a/cylc/flow/task_pool.py b/cylc/flow/task_pool.py
index 92905becf..a71e8bc9e 100644
--- a/cylc/flow/task_pool.py
+++ b/cylc/flow/task_pool.py
@@ -1755,6 +1755,7 @@ class TaskPool:
 
     def sim_time_check(self, message_queue: 'Queue[TaskMsg]') -> bool:
         """Simulation mode: simulate task run times and set states."""
+        print('%')
         if not self.config.run_mode('simulation'):
             return False
         sim_task_state_changed = False
@@ -1783,6 +1784,8 @@ class TaskPool:
                 else:
                     # Simulate message outputs.
                     for msg in itask.tdef.rtconfig['outputs'].values():
+                        # this line doesn't seem to get called by default
+                        print('#', msg)
                         message_queue.put(
                             TaskMsg(job_d, now_str, 'DEBUG', msg)
                         )
```

Author: Counter-example :

```
[scheduler]
    allow implicit tasks = true

[scheduling]
    [[graph]]
        R1 = """
            foo:succeed => after_success
            foo:bar  => bar_follower
            foo:qux  => qux_follower
        """

[runtime]
    [[foo]]
        script = false
        [[[outputs]]]
            bar = 'It\'s not Irrelephant'
            qux = 'It\'s a Hippopotamus'
```

leads to the following:

```
INFO - [1/foo running job:01 flows:1] => succeeded
INFO - [1/after_success waiting(runahead) job:00 flows:1] => waiting
INFO - [1/after_success waiting job:00 flows:1] => waiting(queued)
INFO - [1/bar_follower waiting(queued) job:00 flows:1] => waiting
INFO - [1/qux_follower waiting(queued) job:00 flows:1] => waiting
INFO - [1/after_success waiting(queued) job:00 flows:1] => waiting
```
... then they run

I tried putting the print statement in as a `LOG.critical(msg)` where you had a print statement and got:

```
CRITICAL - It\'s not Irrelephant
CRITICAL - It\'s a Hippopotamus

```

Reviewer: Custom outputs :vomiting_face:. That's really nasty for branched workflows where it could activate multiple branches for custom output switches, but not for succeeded/failed switches!

Counter counter example:

```
[scheduler]
    allow implicit tasks = true

[scheduling]
    [[graph]]
        R1 = """
            foo:started => started
            foo:submitted? => submitted
            foo:submit-failed? => submitted
            foo:expired => expired
            foo:succeed? => succeed
            foo:failed? => failed
            foo:custom  => custom
        """

[runtime]
    [[foo]]
        script = false
        [[[outputs]]]
            custom = custom
```

```
$ cylc vip . -N --mode=simulation 2>&1 | grep '=> waiting(queued)'
2023-10-16T14:10:12+01:00 INFO - [1/foo waiting job:00 flows:1] => waiting(queued)
2023-10-16T14:10:23+01:00 INFO - [1/custom waiting job:00 flows:1] => waiting(queued)
2023-10-16T14:10:23+01:00 INFO - [1/succeed waiting job:00 flows:1] => waiting(queued)
```

So in this case the only outputs produced were the succeeded and the custom but NOT:

* submitted
* submit-failed
* started
* expired
* failed

Author: OK. I'll go back and examine the logic in detail.

Reviewer: From the above I don't think that's right either?

Author: I'm going to create a spin off issue for a possible bugfix, and focus on documenting the actual situation, else this is mission creep:

```suggestion
Tasks always complete all custom outputs.

By default they succeed, but they can be configured to fail using
`[]
```

Author: I've written up the situation as it currently is, and created [this issue](https://github.com/cylc/cylc-flow/issues/5774) - @oliver-sanders can you have a look and say whether you are happy with the current state of play. I suggest not resolving this discussion to make it easier to find using the link on the new issue.

But you may want to test how a workflow handles task failure.

You can set some or all instances of a task to fail using
:cylc.conf:`[runtime][<namespace>][simulation]fail cycle points`.
`fail cycle points` takes either a list of cycle point strings or "all".

Tasks set to fail will succeed on their second or following simulated
submission. If you want all submissions to fail, set

>  :cylc.conf:`[runtime][<namespace>][simulation]fail try 1 only = False`.

Reviewer: ```suggestion
:cylc.conf:`[runtime][<namespace>][simulation]fail try 1 only = False`.
```

run length is computed by dividing the time limit by the speedup factor.

Simulated Failure
^^^^^^^^^^^^^^^^^

> Tasks always complete custom outputs.
> 
> By default they succeed, and return a succeeded output.

Reviewer: ```suggestion
Tasks always complete custom outputs, by default they succeed.
```

.. warning::

   In simulation mode a succeeded output may not imply that
   submitted and/or started outputs are generated - so you will not
   be able to test graph pathways such as ``task:started => do_something``.

You can set some or all instances of a task to fail using
:cylc:conf:`[runtime][<namespace>][simulation]fail cycle points`.
``fail cycle points`` takes either a list of cycle point strings or "all".

Tasks set to fail will succeed on their second or following simulated
submission. If you want all submissions to fail, set
:cylc:conf:`[runtime][<namespace>][simulation]fail try 1 only = False`.

For example, to simulate a task you know to be flaky on the half
hour but not on the hour:

.. code-block:: cylc

   [[get_observations]]
      execution retry delays = PT30S
      [[[simulation]]]
         fail cycle points = 2022-01-01T00:30Z,  2022-01-01T01:30Z

In another case you might not expect the retry to work, and want to test
whether your failure handling works correctly:

.. code-block:: cylc

   [[get_data]]

>        execution retry delays = PT30S

Reviewer: Does this work?

Author: If by "work" you mean "fails multiple times", then it works. Although thinking about it it's a slightly silly example because the `server_broken` output will appear at the same time as the first failure.

>        [[[outputs]]]
>           server_broken = "data server unreachable"

Author: ```suggestion
```

>        [[[outputs]]]
>           server_broken = "data server unreachable"

Author: ```suggestion
```
